File: tests/planner.test.ts
import { buildTodayPlan, scoreTask } from "../src/core/planner";
import type { Task } from "../src/core/types";

function inDays(days: number): string {
  const d = new Date();
  d.setDate(d.getDate() + days);
  return d.toISOString().split("T")[0];
}

describe("scoreTask", () => {
  it("prioriza tareas altas y urgentes", () => {
    const highUrgent: Task = {
      id: "1",
      title: "Enviar propuesta",
      priority: "high",
      dueDate: inDays(0),
      estimatedMinutes: 30,
      completed: false,
    };

    const lowNoDate: Task = {
      id: "2",
      title: "Ordenar archivos",
      priority: "low",
      estimatedMinutes: 30,
      completed: false,
    };

    expect(scoreTask(highUrgent)).toBeGreaterThan(scoreTask(lowNoDate));
  });
});

describe("buildTodayPlan", () => {
  it("devuelve máximo 3 tareas no completadas ordenadas por score", () => {
    const tasks: Task[] = [
      { id: "1", title: "A", priority: "low", estimatedMinutes: 20, completed: false },
      { id: "2", title: "B", priority: "high", dueDate: inDays(1), estimatedMinutes: 40, completed: false },
      { id: "3", title: "C", priority: "medium", dueDate: inDays(0), estimatedMinutes: 25, completed: false },
      { id: "4", title: "D", priority: "high", estimatedMinutes: 200, completed: false },
      { id: "5", title: "E", priority: "high", estimatedMinutes: 20, completed: true }
    ];

    const plan = buildTodayPlan(tasks);

    expect(plan).toHaveLength(3);
    expect(plan.find((task) => task.id === "5")).toBeUndefined();

    for (let i = 0; i < plan.length - 1; i++) {
      expect(scoreTask(plan[i])).toBeGreaterThanOrEqual(scoreTask(plan[i + 1]));
    }
  });
});

File: internal/planner/planner_test.go
package planner

import (
	"testing"
	"time"
)

func TestScoreTaskPrioritizesUrgentHigh(t *testing.T) {
	now := time.Date(2026, 3, 2, 8, 0, 0, 0, time.UTC)
	due := now

	highUrgent := Task{ID: "1", Title: "urgent", Priority: High, DueDate: &due, EstimatedMinutes: 30}
	low := Task{ID: "2", Title: "low", Priority: Low, EstimatedMinutes: 30}

	if ScoreTask(highUrgent, now) <= ScoreTask(low, now) {
		t.Fatal("expected urgent high task to have a bigger score")
	}
}

func TestBuildTodayPlanFiltersAndLimits(t *testing.T) {
	now := time.Date(2026, 3, 2, 8, 0, 0, 0, time.UTC)
	due := now
	tasks := []Task{
		{ID: "1", Priority: Low, EstimatedMinutes: 20},
		{ID: "2", Priority: High, DueDate: &due, EstimatedMinutes: 40},
		{ID: "3", Priority: Medium, EstimatedMinutes: 25},
		{ID: "4", Priority: High, EstimatedMinutes: 200},
		{ID: "5", Priority: High, EstimatedMinutes: 20, Completed: true},
	}

	plan := BuildTodayPlan(tasks, now, 3)
	if len(plan) != 3 {
		t.Fatalf("expected 3 tasks, got %d", len(plan))
	}
	for _, task := range plan {
		if task.ID == "5" {
			t.Fatal("completed task should not be included")
		}
	}
}
